### DISCUSSION TITLE
back/next navigation between survey pages with jspsych

### DISCUSSION THREAD

**Discussion by Noam-Yair** [jspsych/jsPsych#1686] [Q&A]
I'm trying to add a back button to the trial so that if someone didn't answer or wants to change his/her answer, he/she can click the button and see the last question.
If someone can share the implementation of a back button for any kind of trial that would be great, since I would like it to be on a various kinds of trails. 

currently I'm working on a trial that will be something like this:

```html
 <!DOCTYPE html>
<html>
<head>
    <title>My experiment</title>
    <script src="jspsych-6.3.0/jspsych.js"></script>
    <script src="jspsych-6.3.0/plugins/jspsych-html-button-response.js"></script>
    <script src="jspsych-6.3.0/plugins/jspsych-image-keyboard-response.js"></script>
    <link href="jspsych-6.3.0/css/jspsych.css" rel="stylesheet" type="text/css">
</head>
<body></body>
<script>

    var face_name_procedure = {
        timeline: [
            {
                type: 'html-button-response',
                stimulus: function(){
                    var html="<p>"+jsPsych.timelineVariable('sentence')+"</p>";
                    return html;
                },
                choices: ["Pleasant", "Unpleasant"],
                trial_duration: 2500
            },{
                type: 'html-button-response',
                stimulus: function(){
                    var html="<p>"+jsPsych.timelineVariable('numbers')+"</p>";
                    return html;
                },
                choices: ["True", "False"],
                trial_duration: 2500
            }

        ],
        timeline_variables: [
            { sentence: 'sentence1' , numbers: 'number1'},
            { sentence: 'sentence2' , numbers: 'number2'},
            { sentence: 'sentence3' , numbers: 'number3'},
            { sentence: 'sentence4' , numbers: 'number4'}
        ]
    }




    jsPsych.init({
        timeline: [face_name_procedure]
    })

</script>
</html>
```

**Comment by jodeleeuw** [comment]
A true back button isn't possible with jsPsych's current architecture unless it is contained *within* a plugin. For example the `instructions` plugin has a back button implemented. That plugin works by taking a list of all the pages and allowing the participant to navigate through them.

The reason a back button can't be easily added to other plugins is that the underlying code for an experiment's timeline only allows going forward. Even in the case of looping timelines, the data structures treat the loop as a new set of trials and so there is data for each loop through the timeline. Going back to change existing data is not (easily) done. You could probably fake it through some clever functions that pre-filled in data based on previous trial data, but it will be complicated to do because this isn't a feature we've ever intended to cover. 

If this is a feature that you really need then jsPsych might not be the right tool at this point in time.

**Reply by Noam-Yair** [reply]
Thank you so much for the detailed answer! I'll try to make a new plugin as you suggested!
Thanks again

**Comment by berentelg** [comment]
I am interested in this kind of feature as well.

I want to show participants several images and they should be able to go back and forth between those images. After this viewing section they will be asked to rate each image (which can be solved by the "jspsych-image-slider-response" plugin). But I am not quite sure how to solve the first part...

**Reply by becky-gilbert** [reply]
Hi @berentelg, as Josh said in his comment, allowing participants to go back and forth freely among separate trials isn't something that jsPsych isn't really designed to do. One exception to this is the [instructions plugin](https://www.jspsych.org/plugins/jspsych-instructions/), which allows the participant to go back and forth among multiple pages - this works because the whole set of pages is contained within a single 'trial' (plugin). The instructions plugin doesn't work as a solution for the original poster's question, because they wanted to allow participants to go back to previous survey questions in order to view/modify their previous answers. But in your case, since you're just presenting static content (images, rather than participants' previous answers), you could do this using the instructions plugin.

The `pages` parameter in the instructions plugin must be an array of strings, where each string is the HTML that should be presented on that page. Because you're presenting images, you will need to embed the images into the HTML using the `img` tag, e.g.
```js
var image_presentation = {
  type: 'instructions',
  pages: ['<img src="image1.png">', '<img src="image2.png">'],
  show_clickable_nav: true
}
```
Don't forget to preload your images with the [preload plugin](https://www.jspsych.org/plugins/jspsych-preload/). Because you're embedding the image files in HTML strings, you can't rely on jsPsych's `auto_preload` option and will instead need to [preload the files manually](https://www.jspsych.org/overview/media-preloading/#manual-preloading).
I hope this helps. Let us know whether or not this works for you.

**Reply by berentelg** [reply]
I tried this solution just a few minutes before you wrote the comment. So I guess at least your comment provided me some assurance that I am not completely off track.

Anyway,  thanks for your help @becky-gilbert!
